### PR TITLE
add csv builders

### PR DESCRIPTION
**Pull request by hjaensch7**
- csv builders for projects, forests, and project updates
- data backing airtable
- PALS integration

### CODE REVIEW

File: PALS_to_csv.go [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+
+type PALSToCsvConfig struct {
+	PalsDataFile string `required help:"PALS Data CSV File" type:"path"`
+}
+
+func PalsToCsv(config PALSToCsvConfig) error {
+	filePath := "data/pals.csv"
+	err := writePALSCsv(config.PalsDataFile, filePath)
+	if err != nil {
+		log.WithFields(log.Fields{
```
Comment (maloneya) [inline PALS_to_csv.go:19]: Worth considering where in the program you want to log your errors. 

if by default you log an error every time you `if err != nil` you end up with pretty polluted log stream

my suggestion would be to have each abstraction decorate the error with its relevant context, i.e. 
```
return fmt.Errorf("Unable to write CSV %w", err) 
```
and allow the top level controller to do the logging

File: PALS_to_csv.go [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+			"file":  filePath,
+			"error": err.Error(),
+		}).Error("Unable to write CSV")
+		return err
+	}
+	return nil
+
+}
+
+type PalsEntry struct {
```
Comment (maloneya) [inline PALS_to_csv.go:29]: should this be in Types.go?

File: PALS_to_csv.go [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+		palsRow.Forest,
+		palsRow.CalendarYearSigned,
+		palsRow.CalendarYearInitiated,
+		palsRow.OverallCaseOutcome,
+		palsRow.CaseStatus,
+	}
+}
+
+func writePALSCsv(original_path string, path string) error {
+	// 26 - 43
```
Comment (maloneya) [inline PALS_to_csv.go:120]: What does this mean?

File: PALS_to_csv.go [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+	purposes := map[string]string{
```
Comment (maloneya) [inline PALS_to_csv.go:121]: why are these local variables to the function? perhaps should be a global constant, or also stored in types.go?

File: PALS_to_csv.go [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+		"TR": "Travel management",
+		"TS": "Timber sales (green)",
+		"WC": "Watershed improvements",
+		"WD": "Wilderness management",
+		"WI": "Wind",
+	}
+	//open original file
+	f, err := os.Open(original_path)
+	if err != nil {
+		log.Fatal(err)
```
Comment (maloneya) [inline PALS_to_csv.go:196]: Nit pick: `log.Fatal(fmt.Errorf("writePalstoCSV open call err: %w", err))` would make it easier to pinpoint where in the program the error came from

File: PALS_to_csv.go [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+	data, err := csvReader.ReadAll()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var newPals []PalsEntry
+	headers := data[0]
+	for _, line := range data[1:] {
+		// only get projects after 2012
+		if yearAsInt, _ := strconv.Atoi(line[10]); yearAsInt < 2012 {
```
Comment (maloneya) [inline PALS_to_csv.go:209]: Lets store the year on `PALSToCsvConfig` and pass it in from main

File: PALS_to_csv.go [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+			continue
+		}
+		var currentActivities []string
+		var currentPurposes []string
+		// purposes
+		for i, purposeBool := range line[26:43] {
```
Comment (maloneya) [inline PALS_to_csv.go:215]: defining all these int literals as named constants will improve readability

File: PALS_to_csv.go [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+			if purposeBool == "1" {
+				purposeCode := headers[i+26]                                     // FC Facility Management
+				purposeCode = purposeCode[0:2]                                   // FC
+				currentPurposes = append(currentPurposes, purposes[purposeCode]) // Facility Management
+			}
+		}
+		// activities
+		for i, activityBool := range line[44:93] {
```
Comment (maloneya) [inline PALS_to_csv.go:223]: Possible cleaner structure? 

``` 
ACTIVITY_START := 44
ACTIVITY_END := 93
FACTS_ACTVITY := 53
for i := ACTIVITY_START; i < ACTIVITY_END; i++ {
   if i == FACTS_ACTIVITY 
     continue 
    }
   activityCode := headers[i]
   ...
}

File: PALS_to_csv.go [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+			Forest:                            line[102],
+			CalendarYearSigned:                line[103],
+			CalendarYearInitiated:             line[104],
+			OverallCaseOutcome:                line[105],
+			CaseStatus:                        line[106],
+		})
+	}
+
+	// open file
+	csvFile, err := os.Create(path)
```
Comment (maloneya) [inline PALS_to_csv.go:276]: bummer to throw away all of our work if we get some error here. maybe we open all the files first, do the transform then write

File: PALS_to_csv.go [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+	if err != nil {
+		log.Fatalf("failed creating file: %s", err)
+		return err
+	}
+	defer csvFile.Close()
+
+	// build column headers
+	writer := csv.NewWriter(csvFile)
+	writer.Write([]string{
```
Comment (maloneya) [inline PALS_to_csv.go:285]: any way we can leverage the existing headers for this and just skip the activity?

File: forest_json_to_csv.go [hunk 11]
```diff
@@ -67,18 +70,194 @@ func ForestJsonToCsv(config ForestJsonToCsvConfig) error {
 	return nil
 }
 
-// TODO(hank)
 func writeForestCsv(forests []Forest, path string) error {
+	// open file
+	csvFile, err := os.Create(path)
```
Comment (maloneya) [inline forest_json_to_csv.go:75]: just a general call out that there are multiple places in the code base that deal with the file and csv apis, we could make a single utility that was just 
```
func writeCSV(header []string. rows string[][]) error
```
and then each transform is just responsible for produce the new rows, but doesn't care about the CSV details

File: types.go [hunk 12]
```diff
@@ -42,11 +43,12 @@ func (forest Forest) AsCsv() [][]string {
 			project.Contact.Phone,
 			project.Description,
 			project.WebLink,
-			project.Location,
+			project.Location[10:],
```
Comment (maloneya) [inline types.go:46]: ??? should probably explain in a comment

File: types.go [hunk 13]
```diff
@@ -153,3 +155,16 @@ func trim(s string) string {
 	s = singleSpacePattern.ReplaceAllString(s, " ")
 	return s
 }
+
+func getDocsAsString(project ProjectUpdate) string {
+	allDocs, _ := json.Marshal(project.ProjectDocuments)
```
Comment (maloneya) [inline types.go:160]: error?

File: forest_json_to_csv.go [hunk 14]
```diff
@@ -67,18 +70,194 @@ func ForestJsonToCsv(config ForestJsonToCsvConfig) error {
...
+	}
+
+	// iterate through leftovers and add states
+	for _, forest := range leftoverForests {
+		if forestInMap, ok := forestMap[forest.Id]; ok {
+			forestInMap.State = forestInMap.State + ", " + forest.State
+			forestMap[forest.Id] = forestInMap
+		}
+	}
+
```
Comment (maloneya) [inline forest_json_to_csv.go:176]: Maybe we make a separate `Transform` type function that captures all this business logic?

Comment (maloneya) [reply forest_json_to_csv.go:176]: i would also say that this code is pretty hard to follow

**Review by maloneya — COMMENTED** [review]
Feel free to merge, none of my concerns are blocking. 

tl;dr
1. some suggestions for errors and logging
2. some organizational suggestions with our type and schema definitions
3. some organizational suggestions for csv utilities
4. some style suggestions to improve readability  

Additionally i think that `writeProjectsCsv` is currently super confusing, i don't really understand how the dedupe and `leftoverForests` things is supposed to work. There is likely a much simpler way of accomplishing this transformation 

i would also suggest that we just store the CSVs in Drive not Github
